homeassistant/components/marstek/__init__.py

```diff
@@ -0,0 +1,84 @@
+"""The Marstek integration."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+
+from pymarstek import MarstekUDPClient, get_es_mode
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DEFAULT_UDP_PORT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["sensor"]
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+
+@dataclass
+class MarstekRuntimeData:
+    """Runtime data for Marstek config entries."""
+
+    udp_client: MarstekUDPClient
+
+
+type MarstekConfigEntry = ConfigEntry[MarstekRuntimeData]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Marstek component."""
+    store = hass.data.setdefault(DOMAIN, {})
+    if "udp_client" not in store:
+        client = MarstekUDPClient()
+        await client.async_setup()
+        store["udp_client"] = client
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: MarstekConfigEntry) -> bool:
+    """Set up Marstek from a config entry."""
+    _LOGGER.info("Setting up Marstek config entry: %s", entry.title)
+
+    store = hass.data.setdefault(DOMAIN, {})
```

Comment:
You only need to use `entry.runtime_data`, there is no need to store anything in `hass.data` for this.